File: flutter/lib/desktop/plugin/desc.dart
import 'dart:collection';

const String kValueTrue = '1';
const String kValueFalse = '0';

class UiType {
  String key;
  String text;
  String tooltip;
  String action;

  UiType(this.key, this.text, this.tooltip, this.action);

  UiType.fromJson(Map<String, dynamic> json)
      : key = json['key'] ?? '',
        text = json['text'] ?? '',
        tooltip = json['tooltip'] ?? '',
        action = json['action'] ?? '';

  static UiType? create(Map<String, dynamic> json) {
    if (json['t'] == 'Button') {
      return UiButton.fromJson(json['c']);
    } else if (json['t'] == 'Checkbox') {
      return UiCheckbox.fromJson(json['c']);
    } else {
      return null;
    }
  }
}

class UiButton extends UiType {
  String icon;

  UiButton(
      {required String key,
      required String text,
      required this.icon,
      required String tooltip,
      required String action})
      : super(key, text, tooltip, action);

  UiButton.fromJson(Map<String, dynamic> json)
      : icon = json['icon'] ?? '',
        super.fromJson(json);
}

class UiCheckbox extends UiType {
  UiCheckbox(
      {required String key,
      required String text,
      required String tooltip,
      required String action})
      : super(key, text, tooltip, action);

  UiCheckbox.fromJson(Map<String, dynamic> json) : super.fromJson(json);
}

class Location {
  // location key:
  //  host|main|settings|display|others
  //  client|remote|toolbar|display
  HashMap<String, UiType> ui;

  Location(this.ui);
  Location.fromJson(Map<String, dynamic> json) : ui = HashMap() {
    json.forEach((key, value) {
      var ui = UiType.create(value);
      if (ui != null) {
        this.ui[ui.key] = ui;
      }
    });
  }
}

class ConfigItem {
  String key;
  String value;
  String description;
  String defaultValue;

  ConfigItem(this.key, this.value, this.defaultValue, this.description);
  ConfigItem.fromJson(Map<String, dynamic> json)
      : key = json['key'] ?? '',
        value = json['value'] ?? '',
        description = json['description'] ?? '',
        defaultValue = json['default'] ?? '';

  static String get trueValue => kValueTrue;
  static String get falseValue => kValueFalse;
  static bool isTrue(String value) => value == kValueTrue;
  static bool isFalse(String value) => value == kValueFalse;
}

class Config {
  List<ConfigItem> local;
  List<ConfigItem> peer;

  Config(this.local, this.peer);
  Config.fromJson(Map<String, dynamic> json)
      : local = (json['local'] as List<dynamic>)
            .map((e) => ConfigItem.fromJson(e))
            .toList(),
        peer = (json['peer'] as List<dynamic>)
            .map((e) => ConfigItem.fromJson(e))
            .toList();
}

class Desc {
  String id;
  String name;
  String version;
  String description;
  String author;
  String home;
  String license;
  String published;
  String released;
  String github;
  Location location;
  Config config;

  Desc(
      this.id,
      this.name,
      this.version,
      this.description,
      this.author,
      this.home,
      this.license,
      this.published,
      this.released,
      this.github,
      this.location,
      this.config);

  Desc.fromJson(Map<String, dynamic> json)
      : id = json['id'] ?? '',
        name = json['name'] ?? '',
        version = json['version'] ?? '',
        description = json['description'] ?? '',
        author = json['author'] ?? '',
        home = json['home'] ?? '',
        license = json['license'] ?? '',
        published = json['published'] ?? '',
        released = json['released'] ?? '',
        github = json['github'] ?? '',
        location = Location(HashMap<String, UiType>.from(json['location'])),
        config = Config(
            (json['config'] as List<dynamic>)
                .map((e) => ConfigItem.fromJson(e))
                .toList(),
            (json['config'] as List<dynamic>)
                .map((e) => ConfigItem.fromJson(e))
                .toList());
}

final mapPluginDesc = <String, Desc>{};

void updateDesc(Map<String, dynamic> desc) {
  Desc d = Desc.fromJson(desc);
  mapPluginDesc[d.id] = d;
}

Desc? getDesc(String id) {
  return mapPluginDesc[id];
}

File: src/plugin/desc.rs
use hbb_common::ResultType;
use serde_derive::{Deserialize, Serialize};
use serde_json;
use std::collections::HashMap;
use std::ffi::{c_char, CStr};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UiButton {
    key: String,
    text: String,
    icon: String, // icon can be int in flutter, but string in other ui framework. And it is flexible to use string.
    tooltip: String,
    action: String, // The action to be triggered when the button is clicked.
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UiCheckbox {
    key: String,
    text: String,
    tooltip: String,
    action: String, // The action to be triggered when the checkbox is checked or unchecked.
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
pub enum UiType {
    Button(UiButton),
    Checkbox(UiCheckbox),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Location {
    pub ui: HashMap<String, Vec<UiType>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfigItem {
    pub key: String,
    pub default: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config {
    pub shared: Vec<ConfigItem>,
    pub peer: Vec<ConfigItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishInfo {
    pub published: String,
    pub last_released: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Meta {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    #[serde(default)]
    pub platforms: String,
    pub author: String,
    pub home: String,
    pub license: String,
    pub source: String,
    pub publish_info: PublishInfo,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Desc {
    meta: Meta,
    need_reboot: bool,
    location: Location,
    config: Config,
    listen_events: Vec<String>,
}

impl Desc {
    pub fn from_cstr(s: *const c_char) -> ResultType<Self> {
        let s = unsafe { CStr::from_ptr(s) };
        Ok(serde_json::from_str(s.to_str()?)?)
    }

    pub fn meta(&self) -> &Meta {
        &self.meta
    }

    pub fn location(&self) -> &Location {
        &self.location
    }

    pub fn config(&self) -> &Config {
        &self.config
    }

    pub fn listen_events(&self) -> &Vec<String> {
        &self.listen_events
    }
}
